### DISCUSSION TITLE
Mac Plus access time is consistently 13 ms

### DISCUSSION THREAD

**Discussion by liamhays** [erichelgeson/BlueSCSI-v1#179] [Q&A]
Hello,
I have a DB25 BlueSCSI from tom.scsi.blue, and while it works wonderfully with my Mac Plus, testing with Transoft SCSI Pro 4 finds an access time of consistently about 13 milliseconds. I did some research, and it seems that most quality SD cards have access times of 1 or 2 ms. I have tested both a 32 GB Sandisk Ultra and a 256 GB Samsung EVO Select, both formatted as exFAT, and gotten the same results.

I have a project that needs low access times, because it needs to read small amounts of data from the disk quickly. The "seek time" given by SCSI Director is about 2 ms (totally acceptable), and the read rate is about 430 kB/s. I'm using the `r0w0s6` version of SCSI Accelerator. 

I dug into the BlueSCSI code and it looks like all file reading is passed through SdFat, so I don't think the slowdown is caused by an issue there. SdFat is a reputable and high-speed library, even over SPI. The cards are also run at 50 MHz consistently. 

The Retro Theory Twitter post shows access times of about 1.5 ms, and access time *seems* like it should be disk-limited, not computer-limited. Is there a solution? Or is this just a limitation of the early SCSI implementation in the Plus?

**Comment by erichelgeson** [comment]
I'm dong some testing on my Plus for some other things so I'll let you know what I see too, but my initial thought is this is a 8mhz 68000, and this speed/access time is about what you'll expect. Once you get past the SE though seek goes down to sub 1ms.

I would validate what SCSI drivers your using (and try some other ones such as Silver Lining or Apples) to see if they make a difference. Also make sure that SD card is formatted to exFat and do a full overwrite format with SD Card Utility (linked in the Troubleshooting wiki page)

**Reply by liamhays** [reply]
I'm using the HDA blanks from the wiki, and I assume they come with the Apple driver. I've tested with the Transoft driver (via "Install" in Transoft Pro) and the seek time was the same, but the read/write speed were cut in half. I might test it again, though, and I haven't tried Silver Lining yet.

I agree that a 8 MHz 68k probably won't have instantaneous disk seek. However, I don't see why it should be taking ~78,000 cycles either. Where is the time going? I would think that there's a master table or something in RAM that the computer uses to seek to various points on the disk, and accessing RAM and issuing a SCSI command should be pretty fast. I don't know, though.

**Reply by liamhays** [reply]
Ok! Some more progress. All drivers have the same seek time (down to a couple hundredths of a millisecond), except for Butterfly, where the Apple driver is 2 ms faster. The Transoft driver has a very slow data rate, about 160 kB/s read/write, while the Apple and Silverlining drivers have identical performance at about 430 kB/s read/write.

The `SCSI Results` file created by Transoft says that the measured seek time is the time needed to receive a SCSI seek command complete message. I'm thinking there's two possible scenarios here:
* The BlueSCSI is delaying the seek complete command---I highly doubt this, the source code appears to be expertly written
* Because the advantages provided by SCSI Accelerator don't apply to commands, only blind transfers, the Mac waits far longer than it needs to before it reads the bus

Finally and unsurprisingly, enabling disk cache does not affect seek time or performance.

**Reply by erichelgeson** [reply]
> Because the advantages provided by SCSI Accelerator don't apply to commands, only blind transfers, the Mac waits far longer than it needs to before it reads the bus

This is an interesting theory and might be the case. If you notice during the speed test the seek is minimal (~2-3ms).

**Reply by liamhays** [reply]
I tried getting the Plus to run System 7. I do not recommend :). Seriously, though, it booted maybe half the time, and threw a system error the other half. Transoft Pro 4 crashes on "Test" with an illegal instruction (even though it works on System 6), and FWB HDT Benchtest ran but hung on the first test.

I was interested in System 7.5 because it (or a nearby version) includes SCSI Manager 4.3, which is a much improved version of the SCSI Manager in ROM. Although seeking should be down to the driver, not the ROM code, I thought it was worth a shot. However, System 7 is pain to get working. If somebody has an SE or a Classic (both 8MHz 68000, but different ROMs), could you see if performance changes under System 7.5?
